### DISCUSSION TITLE
How to add additional custom SyntaxHighlighter's

### DISCUSSION THREAD

**Discussion by ghandic** [swagger-api/swagger-ui#7551] [Q&A]
Using this reference to implement different languages on the snippets https://github.com/swagger-api/swagger-ui/pull/7181#issuecomment-919617381

I can get other languages like Python... However, I don't get any syntax highlighting. Is there a way to allow this? I'm putting together a PR to include options for FastAPI

<img width="689" alt="Screenshot 2021-10-11 at 14 49 54" src="https://user-images.githubusercontent.com/23500353/136730683-e44e0406-3a77-4159-9a39-45922c5e97f4.png">

This is where all of the languages are added https://github.com/swagger-api/swagger-ui/blob/cc408812fc927e265da158bf68239530740ab4cc/src/core/syntax-highlighting.js

**Comment by mathis-m** [comment]
@ghandic Actually I saw this comming when I submitted the RequestSnippets PR back then. This is currently not supported but can be implemented.
@char0n Shall I prepare a PR to provide `SyntaxHighlighter` to the plugin system so other languages can be registered later on?

**Reply by char0n** [reply]
@mathis-m sure go ahead. As you mentioned we need to provide mechanism for plugins to intercept `SyntaxHighlighter` and register new languages. 

Maybe I'm thing ahead of myself here, but can we imagine a usecase where somebody wants to replace syntax highlight mechanism with his own?

**Reply by mathis-m** [reply]
@char0n could be possible. I would recommend we just register the component to the plugin system. This way it can be wrapped or replaced like any other component! With the wrapComponent it should be possible to register more languages.

**Reply by ghandic** [reply]
Would that allow us to use prism too? We use the bundle.js in a raw html script tag currently so would this plug-in system work without the need for any extra css etc that is needed at runtime? 

Excuse my lack of understanding of what registering a language does and how the build of the package works

**Reply by mathis-m** [reply]
@ghandic if prism can be used I'll check for you tomorrow, but adding a other language would require you to write a small plugin(I will document it and post a exact example for your case here) where you add the wanted language to swagger ui. You can for example use unpkg to load the language highlight via script tag into your html and pass it to the plugin. I'll comeback to you when my PR is finished up!

**Reply by ghandic** [reply]
That would be awesome thank you! Appreciate how quick you guys were on this

**Reply by ghandic** [reply]
fyi this is the proposed implementation we have for FastAPI - I'm adding in plugin support. Would love to also add your date plugin @mathis-m if you could help me understand how to use it with the swagger-ui-bundle.js

```python
def get_swagger_ui_html(
    *,
    openapi_url: str,
    title: str,
    swagger_js_url: str = "https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui-bundle.js",
    swagger_css_url: str = "https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui.css",
    swagger_favicon_url: str = "https://fastapi.tiangolo.com/img/favicon.png",
    oauth2_redirect_url: Optional[str] = None,
    init_oauth: Optional[Dict[str, Any]] = None,
    plugins: Optional[List[Plugin]] = None
) -> HTMLResponse:

    if plugins is None:
        plugins = []

    html = f"""
    <!DOCTYPE html>
    <html>
        <head>
            <link type="text/css" rel="stylesheet" href="{swagger_css_url}">
            <link rel="shortcut icon" href="{swagger_favicon_url}">
            <title>{title}</title>
        </head>
        <body>
            <div id="swagger-ui"></div>
            <script src="{swagger_js_url}"></script>
            {"".join("".join(f'<script src="{js}"></script>' for js in plugin.js) for plugin in plugins if hasattr(plugin, "js"))}
            <!-- `SwaggerUIBundle` is now available on the page -->
            <script>

            {";".join(plugin.get() for plugin in plugins)}
            
            const ui = SwaggerUIBundle({{
                url: '{openapi_url}',
                dom_id: '#swagger-ui',
                presets: [
                SwaggerUIBundle.presets.apis,
                SwaggerUIBundle.SwaggerUIStandalonePreset
                ],
                layout: "BaseLayout",
                deepLinking: true,
                showExtensions: true,
                showCommonExtensions: true,
                {f"oauth2RedirectUrl: window.location.origin + '{oauth2_redirect_url}'," if oauth2_redirect_url else ""}
                plugins: [
                    SwaggerUIBundle.plugins.DownloadUrl,
                    {",".join(plugin.use() for plugin in plugins)}
                ],
                {",".join(f"...{plugin.config()}" for plugin in plugins)}
            }})
            
            {f"ui.initOAuth({json.dumps(jsonable_encoder(init_oauth))})" if init_oauth else ""}

            </script>
        </body>
    </html>
    """
    return HTMLResponse(html)

```

**Reply by ghandic** [reply]
With the snippetplugin looking something like this


```python
from typing import Dict, List

from pydantic import BaseModel

supported_snippet_languages = {"c_libcurl",
 "csharp_restsharp",
 "go_native",
 "java_okhttp",
 "java_unires",
 "javascript_jquery",
 "javascript_xhr",
 "node_native",
 "node_request",
 "node_unirest",
 "objc_nsurlsession",
 "ocaml_cohttp",
 "php_curl",
 "php_http1",
 "php_http2",
 "python_python3",
 "python_requests",
 "ruby_native",
 "shell_curl",
 "shell_httpie",
 "shell_wget",
 "swift_nsurlsession"}

default_openapi_snippets = {"curl_bash", "curl_powershell", "curl_cmd"}




class Snippet(BaseModel):
    title: str
    syntax: str


class PremadeSnippets:

    _templates = {
        "curl": {"curl_bash": Snippet(title="cURL", syntax="bash")},
        "python": {"python_requests": Snippet(title="* Python - requests", syntax="python")},
        "csharp": {"csharp_restsharp": Snippet(title="* C# - restsharp", syntax="csharp")},
        "go": {"go_native": Snippet(title="* Go", syntax="go")},
        "java": {"java_okhttp": Snippet(title="* Java - okhttp", syntax="java")},
        "jquery": {"javascript_jquery": Snippet(title="* JQuery", syntax="javascript")},
        "node": {"node_native": Snippet(title="* Node - native", syntax="javascript")},
        "objectivec": {"objc_nsurlsession": Snippet(title="* Objective C - NSURLSession", syntax="objectivec")},
        "ocaml": {"ocaml_cohttp": Snippet(title="* OCaml - cohttp", syntax="ocaml")},
        "php": {"php_curl": Snippet(title="* PHP - curl", syntax="php")},
        "ruby": {"ruby_native": Snippet(title="* Ruby - native", syntax="ruby")},
        "swift": {"swift_nsurlsession": Snippet(title="* Swift - NSURLSession", syntax="swift")},
    }

    @staticmethod
    def get(programs:List[str])->Dict[str, Snippet]:
        template = {}
        for p in sorted(programs, key=lambda s: s in default_openapi_snippets):
            template = {**template, **PremadeSnippets._templates.get(p)}
        return template
    
    @staticmethod
    def get_all()->Dict[str, Snippet]:
        template = {}
        for p in sorted(PremadeSnippets._templates.keys(), key=lambda s: s in default_openapi_snippets):
            template = {**template, **PremadeSnippets._templates.get(p)}
        return template



class SnippetPlugin(BaseModel):
    """[summary]

    Kudos for the JS library and suggestions goes to:
    - https://github.com/ErikWittern/openapi-snippet
    - https://github.com/swagger-api/swagger-ui/pull/7181

    Args:
        BaseModel ([type]): [description]

    Returns:
        [type]: [description]
    """
    snippets: Dict[str, Snippet] = PremadeSnippets.get(["curl", "python"])
    js: List = ["/public/openapisnippet.min.js"]

    def __init__(self, **kwargs):
        if "snippets" in kwargs:
            assert supported_snippet_languages.union(default_openapi_snippets).issuperset(set(kwargs["snippets"].keys()))
            kwargs["snippets"]
        super().__init__(**kwargs)



    def config(self)->str:
        generators = self.dict()["snippets"]
        return f"""{{
            requestSnippetsEnabled: true,
            requestSnippets: {{
                generators: {generators},
                languages: {list(self.snippets.keys())}
            }}
        }}
        """

    def use(self)->str:
        return self.__class__.__name__

    def get(self) ->str:
        return f"""
            function fn_generator(target) {{
                    return (req) => {{
                        const {{ spec, oasPathMethod }} = req.toJS();
                        const {{ path, method }} = oasPathMethod;

                        if (typeof spec.servers == "undefined") {{
                            spec.servers = [{{url: window.location.origin}}]
                        }}

                        let snippet;
                        try {{
                            snippet = OpenAPISnippets.getEndpointSnippets(spec, path, method, [target]).snippets[0]
                                .content;
                        }} catch (err) {{
                            console.log(err);
                            snippet = "Unable to convert, try copying the cURL command into https://curl.trillworks.com and selecting your desired language"
                        }}
                        return snippet;
                    }};
                }}

            const {self.__class__.__name__} = {{
                statePlugins: {{
                    spec: {{
                        wrapSelectors: {{
                            requestFor: (ori, system) => (state, path, method) => {{
                                return ori(path, method)
                                    ?.set("spec", state.get("json", {{}}))
                                    ?.setIn(["oasPathMethod", "path"], path)
                                    ?.setIn(["oasPathMethod", "method"], method);
                            }},
                            mutatedRequestFor: (ori) => (state, path, method) => {{
                                return ori(path, method)
                                    ?.set("spec", state.get("json", {{}}))
                                    ?.setIn(["oasPathMethod", "path"], path)
                                    ?.setIn(["oasPathMethod", "method"], method);
                            }},
                        }},
                    }},
                }},
                fn: Object.fromEntries({[s for s in self.snippets.keys() if s not in default_openapi_snippets]}.map((target) => [`requestSnippetGenerator_${{target}}`, fn_generator(target)])),

            }};
        """
```

It then produces the HTML with the plugin and can be used in FastAPI like:

```python
app = FastAPI(openapi_plugins=[SnippetPlugin(snippets=PremadeSnippets.get_all())],)

@app.post("/b")
def do_a(demo: str = Query(..., example="lol"), ex: str = Body(..., example="hey")):
    return {"answer": "a"}
```

Note just using all snippets for example.

**Reply by ghandic** [reply]
@mathis-m I also found that the openapisnippet fails for things like file upload. Any idea on that? ATM I have have just added a message for then to convert themselves by using the CURL to XXX ie `"Unable to convert, try copying the cURL command into https://curl.trillworks.com and selecting your desired language"`

**Reply by ghandic** [reply]
Is there anything I can help with on this?

**Reply by rsanvatsarkar** [reply]
@ghandic ,
Did you manage to get the other languages code snippets support working?
I tried adding "requestSnippets" generator but no luck. It only display the default cURL snippets. I would like to add the C#, JavaScript and Python language support. Please note that, I have take the latest css and the .js (swagger-ui-bundle, swagger-ui-standalone-preset). Any help would be greatly appreciated. Thank you.

const ui = SwaggerUIBundle({
                urls:
        @(Html.Raw(JsonConvert.SerializeObject(_servicesConfig.Services.Select(c => new { url = c.Url.ToString(), name = c.Name }), Formatting.Indented))),
                "urls.primaryName": "@(_servicesConfig.Services.SingleOrDefault(c => c.Default)?.Name ?? _servicesConfig.Services.FirstOrDefault()?.Name)",
                dom_id: '#swagger-ui',
                deepLinking: true,
                requestInterceptor: function (req)
                {
                    req.mode = 'cors';
                    return req
                },
                validatorUrl: null,
                presets: [
                    SwaggerUIBundle.presets.apis,
                    SwaggerUIStandalonePreset
                ],
                plugins: [
                    SwaggerUIBundle.plugins.DownloadUrl
                ],
                layout: "StandaloneLayout",
                defaultModelsExpandDepth: -1,
                requestSnippetsEnabled: true,
                requestSnippets: {
                    generators: {
                        curl_bash: {
                            title: "cURL (bash)",
                            syntax: "bash"
                        },
                        curl_powershell: {
                            title: "cURL (PowerShell)",
                            syntax: "powershell"
                        },
                        curl_cmd: {
                            title: "cURL (CMD)",
                            syntax: "bash"
                        },
                        csharp_restsharp: {
                            title: "* C# - restsharp",
                            syntax: "csharp"
                        },
                        "node_native": {
                            title: "NodeJs Native",
                            syntax: "javascript"
                        }
                    },
                    defaultExpanded: true,
                    languages: ["curl_bash", "curl_powershell", "curl_cmd", "csharp_restsharp", "node_native]
                },
                showExtensions: true,
                showCommonExtensions: true
            })

**Reply by ghandic** [reply]
Please format your code, did you try using the above examples?

**Reply by rsanvatsarkar** [reply]
@ghandic , the code that I have pasted above is my code which is not working. I am once again attaching my code below. Please note that. I have taken the most recent swagger-ui-bundle.js, swagger-ui-standalone-preset.js and the CSS files but it doesn't work. I always see the default cURL code snippets and I am looking to add C#, JavaScript, Python and Ruby code snippets.

const ui = SwaggerUIBundle({
urls:
@(Html.Raw(JsonConvert.SerializeObject(_servicesConfig.Services.Select(c => new { url = c.Url.ToString(), name = c.Name }), Formatting.Indented))),
"urls.primaryName": "@(_servicesConfig.Services.SingleOrDefault(c => c.Default)?.Name ?? _servicesConfig.Services.FirstOrDefault()?.Name)",
dom_id: '#swagger-ui',
deepLinking: true,
requestInterceptor: function (req)
{
req.mode = 'cors';
return req
},
validatorUrl: null,
presets: [
SwaggerUIBundle.presets.apis,
SwaggerUIStandalonePreset
],
plugins: [
SwaggerUIBundle.plugins.DownloadUrl
],
layout: "StandaloneLayout",
defaultModelsExpandDepth: -1,
requestSnippetsEnabled: true,
requestSnippets: {
generators: {
curl_bash: {
title: "cURL (bash)",
syntax: "bash"
},
curl_powershell: {
title: "cURL (PowerShell)",
syntax: "powershell"
},
curl_cmd: {
title: "cURL (CMD)",
syntax: "bash"
},
csharp_restsharp: {
title: "* C# - restsharp",
syntax: "csharp"
},
"node_native": {
title: "NodeJs Native",
syntax: "javascript"
}
},
defaultExpanded: true,
languages: ["curl_bash", "curl_powershell", "curl_cmd", "csharp_restsharp", "node_native]
},
showExtensions: true,
showCommonExtensions: true
})

**Reply by mikenerone** [reply]
@ghandic Did your proposed changes (or something analogous to them to your knowledge) ever make it into FastAPI? I'd love to have the other language snippets available, but not strongly enough that I want to add code/maintenance into my project for it. I'm looking for something in a more ready-to-use form.
